### PR TITLE
WT-6185 Insert full updates into the history store in some corner cases and add a history store c test

### PR DESCRIPTION
**Pull request by keithbostic**
@agorrod, this is a branch that I think reproduces the failure I'm seeing from format in WT-6185.

The early commits (a79da63, 3b930f9, 8cabe24, ca356d3, 14a7998), are all format changes I made around the operations tracing code so I could track this problem better.

The final commit (a9c0bfe) is the reproducer. It's a test program in `test/csuite/wt6185_modify_ts`.

The trick is a hack into `cursor.reset` that forcibly evicts the page (see `__cursor_reset()`). I plan to keep that hack, just not sure how yet, right now it just sets a variable in the `WT_CURSOR_BTREE` structure.

If you run the test program with `-e` (don't do eviction), it runs, if you run it without `-e`, it fails immediately.

### CODE REVIEW

**Comment by sulabhM** [comment]
@keithbostic , I merged develop in this branch since WT-6347 got reverted.

I looked at the program and thought we need the values to remain null-terminated, since the value format is "S" and we do strcmp(). Please revert the commit if I am not fixing a bug.

**Comment by keithbostic** [comment]
@sulabhM 
> I looked at the program and thought we need the values to remain null-terminated, since the value format is "S" and we do strcmp(). Please revert the commit if I am not fixing a bug.

I believe the string is always nul-terminated: there are two nul's in the string, one at the end, set after the `memset`, and one introduced by the `sprintf` call, and the macro only clears the first one. I like having the value buffer initialized to `.` characters because it makes it easier to see what changed. Let me know if you disagree or I'm just missing something!

Thanks for fixing the swapped expected/found reports, too.

**Comment by keithbostic** [comment]
@sulabhM, here's a set of seed values that fail fast:
```
seed 4069248439076046261 
seed 1653486788014654281 
seed 1693967443199398970 
seed 1779673949402730967 
seed 2115099655369014718 
seed 2458345513235624456 
seed 3174248082686941088 
seed 3268906030881661575 
seed 3638308157982622670 
seed 3650533065133994571 
seed 5582057777891596745 
seed 4119284141967397113 
seed 6062009287244467947 
seed 9923920940377255439 
seed 2796510557933869401 
seed 2854878965931603739 
seed 3595137053678658538 
seed 5513616288602927773 
seed 7035533035538247242 
seed 4948996138232857492 
```

**Comment by sulabhM** [comment]
@keithbostic , you are good, I did not notice that there are two nulls in the value strings.

**Comment by quchenhao** [comment]
@keithbostic With my fix, the test is now passing but the code becomes more complicated :(

**Comment by quchenhao** [comment]
@keithbostic I see a test failure in the evergreen

```
[2020/06/04 03:40:28.446] test_wt6185_modify_ts: FAILED: main/351: write(STDIN_FILENO, ".", 1) == 1
 [2020/06/04 03:40:28.446] process aborting
 [2020/06/04 03:40:28.447] Aborted (core dumped)
```

I don't think that is related to my fix. Can you take over from here?

**Comment by quchenhao** [comment]
@agorrod @bvpvamsikrishna This is ready for review.

**Comment by quchenhao** [comment]
@agorrod We don't skip calculating the full value even if updates are squashed so that part is OK. I'll double check if we can simplify the algorithm further given next week.

File: src/history/hs.c [hunk 1]
```diff
@@ -546,22 +546,45 @@ __hs_insert_record(WT_SESSION_IMPL *session, WT_CURSOR *cursor, WT_BTREE *btree,
...
+  WT_UPDATE **updp)
 {
-    if (upd->type == WT_UPDATE_MODIFY) {
-        WT_RET(__wt_buf_set(session, full_value, base_full_value, size));
+    WT_UPDATE *upd;
+    *updp = NULL;
+    __wt_modify_vector_pop(modifies, &upd);
+    if (upd->type == WT_UPDATE_TOMBSTONE) {
+        WT_ASSERT(session, modifies->size > 0);
+        if (upd->start_ts == WT_TS_NONE) {
```
Comment (quchenhao) [inline src/history/hs.c:562]: I'll move this part of the code outside this function in WT-6347.

File: src/history/hs.c [hunk 2]
```diff
@@ -776,68 +778,56 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
...
-                    /* Flag the update as now in the history store. */
-                    F_SET(upd, WT_UPDATE_HS);
-                    ++insert_cnt;
-                    if (squashed) {
-                        WT_STAT_CONN_INCR(session, cache_hs_write_squash);
-                        squashed = false;
-                    }
-                }
-            } else
+            if (upd->start_ts == prev_upd->start_ts && upd->txnid == prev_upd->txnid) {
```
Comment (quchenhao) [inline src/history/hs.c:789]: In future if we want to insert reverse modifies for squashed transactions, we can implement the logic in `__hs_next_upd_full_value `

**Comment by quchenhao** [comment]
@agorrod I'd like to merge this. Can you have a look?

**Comment by tammybailey** [comment]
@quchenhao could you create an Evergreen patch for this with stress testing, please? You'll need to merge the latest changes in develop to get the updated `evergreen.yml`.

**Comment by quchenhao** [comment]
@tammybailey patch build
https://evergreen.mongodb.com/version/5edeebbca4cf4722a096c30d

File: src/history/hs.c [hunk 3]
```diff
@@ -691,67 +711,49 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
...
+        for (prev_upd = NULL; upd != NULL; prev_upd = upd, upd = upd->next) {
             if (upd->txnid == WT_TXN_ABORTED)
                 continue;
+
             WT_ERR(__wt_modify_vector_push(&modifies, upd));
 
             /*
-             * If the update is the second update older than the prepared update and we haven't seen
-             * a tombstone. Mark the update.
+             * Only insert full update to the history store if we write a prepared update to the
```
Comment (agorrod) [inline src/history/hs.c:721]: "Only insert full update" reads better as "Always insert full updates". For this and the two following comments as well.

File: src/history/hs.c [hunk 4]
```diff
@@ -691,67 +711,49 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
...
-                    upd_count = 0;
-                    track_prepare = false;
-                }
-            }
+            if (upd->prepare_state == WT_PREPARE_INPROGRESS)
+                enable_reverse_modify = false;
+
+            /* Only insert full update to the history store if we need to squash the updates. */
+            if (prev_upd != NULL && prev_upd->txnid == upd->txnid &&
+              prev_upd->start_ts == upd->start_ts)
```
Comment (agorrod) [inline src/history/hs.c:729]: There will be cases where a single transaction does multiple updates with different timestamps. That happens on replica set secondaries. Is that OK? i.e: is it safe to save a modify whenever the timestamp differs, even if the transaction ID is the same?

Comment (quchenhao) [reply src/history/hs.c:729]: @agorrod I believe that is OK as long as the timestamps are in order.

**Comment by quchenhao** [comment]
@agorrod MongoDB patch build
https://evergreen.mongodb.com/version/5edf102d850e612665c8fb6e
If tests are OK, I'll push the fix for the comments and merge this change.

File: src/history/hs.c (hunk 4 above)
Comment (agorrod) [reply src/history/hs.c:729]: Sounds good.

**Comment by quchenhao** [comment]
@agorrod The stress tests are passing and the test failures in the mongodb patch build are all known issues. Can you approve the change?

**Comment by agorrod** [comment]
Looks good to me
